lib/kauth/login.go

```diff
@@ -0,0 +1,121 @@
+package kauth
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/astore/common"
+	"github.com/enfabrica/enkit/astore/rpc/auth"
+	"github.com/enfabrica/enkit/lib/kcerts"
+	"github.com/enfabrica/enkit/lib/logger"
+	"github.com/enfabrica/enkit/lib/retry"
+	"github.com/enfabrica/enkit/lib/srand"
+	"github.com/pkg/browser"
+	"golang.org/x/crypto/nacl/box"
+	"golang.org/x/crypto/ssh"
+	"math/rand"
+)
+
+func init() {
+	browser.Stdout = nil
+	browser.Stderr = nil
+}
+
+type EnkitCredentials struct {
+	Token string
+	// The below fields can be possibly empty if the auth server does not support CA certificates.
+	CaHosts        []string
+	CAPublicKey    string
+	PrivateKey     ed25519.PrivateKey
+	SSHCertificate *ssh.Certificate
+}
+
+// PerformLogin will login with the provider auth client, retry and logger. It does not care about the cache.
+// If you wish to save the result, please call SaveCredentials
+func PerformLogin(authClient auth.AuthClient, l logger.Logger, repeater *retry.Options, username, domain string) (*EnkitCredentials, error) {
+	rng := rand.New(srand.Source)
+	pubBox, privBox, err := box.GenerateKey(rng)
+	if err != nil {
+		return nil, err
+	}
+	sshPub, sshPriv, err := kcerts.MakeKeys(kcerts.GenerateED25519)
+	if err != nil {
+		return nil, err
+	}
+	areq := &auth.AuthenticateRequest{
+		Key:    (*pubBox)[:],
+		User:   username,
+		Domain: domain,
+	}
+	l.Infof("Retrieving authentication url.")
+	ares, err := authClient.Authenticate(context.TODO(), areq)
+	if err != nil {
+		return nil, fmt.Errorf("Could not contact the authentication server. Is your connectivity working? Is the server up?\nFor debugging: %w", err)
+	}
+	if username != "" {
```

Comment:
I liked my proposal better of passing a list of functions to "interact with the user". That way, we could do things like disable "opening a browser" from a unit test (does the test open browsers for you? fun?) or fully embed in a non interactive library. I'm ok not doing this in this PR.

Reply:
Ah later, mainly because things like opening from a browser should be more of a fire and forget, and user interaction outside of copy-paste generally is pretty unreliable imo.